lib/Runtime/HostManager/HostManager.cpp

```diff
@@ -92,21 +92,36 @@ HostManager::init(std::vector<std::unique_ptr<DeviceConfig>> configs) {
 
 HostManager::~HostManager() { llvm::toString(clearHost()); }
 
-llvm::Error HostManager::addNetwork(std::unique_ptr<Module> module,
-                                    CompilationContext &cctx,
-                                    bool saturateHost) {
-  std::lock_guard<std::mutex> networkLock(networkLock_);
-  auto functions = module->getFunctions();
-  for (auto &F : functions) {
-    std::string name = F->getName();
+void HostManager::cleanupAddNetwork(llvm::ArrayRef<std::string> names) {
+  for (auto &name : names) {
     auto it = networks_.find(name);
     if (it != networks_.end()) {
-      return MAKE_ERR(GlowErr::ErrorCode::RUNTIME_ERROR,
-                      "Failed to add network: already have a function called " +
-                          name);
+      networks_.erase(it);
     }
   }
+}
 
+llvm::Error HostManager::addNetwork(std::unique_ptr<Module> module,
+                                    CompilationContext &cctx,
+                                    bool saturateHost) {
+  std::vector<std::string> names;
+  {
+    std::lock_guard<std::mutex> networkLock(networkLock_);
+    auto functions = module->getFunctions();
+    for (auto &F : functions) {
+      std::string name = F->getName();
+      auto it = networks_.find(name);
+      if (it != networks_.end()) {
+        cleanupAddNetwork(names);
+        return MAKE_ERR(
+            GlowErr::ErrorCode::RUNTIME_ERROR,
+            "Failed to add network: already have a function called " + name);
+      }
+      // Put an entry into networks to denote this network is being added.
+      networks_[name];
```

Comment:
a bit weird, networks_.insert ?

Reply:
I think you actually want a different map here, `pendingNetworks_` or something - that way you can store the key without needing a dummy value that might accidentally get used.